src/mds/CDir.h

```diff
@@ -198,6 +198,34 @@ class CDir : public MDSCacheObject, public Counter<CDir> {
 protected:
   version_t projected_version;
   mempool::mds_co::list<fnode_t> projected_fnode;
+  int children_under_rstat_propagation;
+
+  bool adjust_subtree_freezing_or_frozen(int x, MDSContext::vec* finished = NULL) {
+    CDentry* dn = NULL;
+    CInode* inode = NULL;
+    bool ret = true;
+
+    do {
+      if (!dn)
+	inode = get_inode();
+      else
+        inode = dn->dir->get_inode();
+      inode->subtree_freezing_or_frozen += x;
+      lgeneric_subdout(g_ceph_context, mds, 20) << __func__
+        << ", adjusting subtree_freezing_or_frozen(x=" << x << "," << inode->subtree_freezing_or_frozen << ") for: " << *inode << dendl;
+      ceph_assert(inode->subtree_freezing_or_frozen >= 0);
+      if (!inode->subtree_freezing_or_frozen && !freeze_tree_state) {
+        ceph_assert(finished);
+        inode->take_waiting(CInode::WAIT_CAN_RSTAT_PROPAGATE, *finished);
+      } else if (x < 0){
+        if (finished)
+          finished->clear();
+        ret = false;
+      }
+      dn = inode->get_projected_parent_dn();
+    } while(dn);
+    return ret;
```

Comment:
why all of these are required.   why 'wait unfreeze' in scatter_nudge() does not work

Reply:
@ukernel Hi, thanks for your review.

I thought there could be the following scenario:

1. a write to a file at /a/b/c/d/f/g/h/I/j/k/test_file happened, and for some reasoner rstat failed to be propagated to /a/b/c/d/f/g/h/I/j/k, so this directory's nestlock would be put into updated_scatterlocks and marked dirty;
2. An "export_dir" happened, and export /a/b/c/d/f/g/h/I to another mds;
3. After the export, /a/b/c/d/f/g/h/I/j/k wouldn't be in the updated_scatterlocks queue or marked dirty, which means there no clew for mds to propagate the changed rstat of /a/b/c/d/f/g/h/I/j/k/test_file

So, I thought the "rstat propagation" and "freeze tree" should be interleaved

Reply:
the new auth mds of /a/b/c/d/f/g/h/I should know its have dirty rstat (dirtied before givent time)

what I like is:

- user uses asok command to ask mds.0 to propagate rstat
- mds.0 asks all mds to propagate rstat dirtied before the time (ask command was issue)
- mds.x periodically report its oldest dirty rstat to mds.0
- mds.0 gather all rstat information, so it know if all rstat dirtied before given time are  propagated

with this approach. we just need to modify Locker::updated_scatterlocks. make it track dirty time.

Reply:
@ukernel Hi, thanks for your advise:-)

Do you mean that we can just make Locker::updated_scatterlocks to track dirty rstat and rely on the current "scatter_tick" to propagate rstat? IIUC, "scatter_tick" happens every "mds_tick_interval" seconds, which means the rstat may go up just one level of directories every "mds_tick_interval" seconds in the worst case. If so, wouldn't it take a very long time to propagate rstat of a very deep subtree?